### PR TITLE
Check changed files before running gradle check

### PR DESCRIPTION
**Pull request by gaiksaya**
### Description
The repository requires the gradle check job status to pass before merging any PR. With `path-ignore` filter, this status check was skipped. However, GitHub does not resolve this conflict or allows to specify the requirement to be optional. See https://github.com/opensearch-project/OpenSearch/pull/13494 still waiting for gradle-check status to be set.
There are multiple discussions on GH community forums https://github.com/orgs/community/discussions/44490 and all of them requires work around as of now. This change adds one more step to gradle-check job to detect the files changed and accordingly decide whether to run the consecutive steps or not. See examples below:
1. Where gradle check runs for change in build.gradle https://github.com/gaiksaya/OpenSearch/pull/10
2. Gradle check steps are skipped for changes in md file https://github.com/gaiksaya/OpenSearch/pull/11

I set my fork requirements to mimic gradle check status as required and this satisfies both conditions

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/gradle-check.yml
Comment (peternied) [inline .github/workflows/gradle-check.yml:36]: We cannot skip files under `.github`, if you wanted to modify this to specific files - code owners, this is OK with me.

File: .github/workflows/gradle-check.yml
Comment (peternied) [inline .github/workflows/gradle-check.yml:35]: I don't trust the `**` annotation, lets be very selective
```suggestion
            release-notes/*.md
```

File: .github/workflows/gradle-check.yml
Comment (peternied) [inline .github/workflows/gradle-check.yml:37]: Lets list the specific files that we want to allow, etc...
```suggestion
            README.md
```

File: .github/workflows/gradle-check.yml
Comment (peternied) [inline .github/workflows/gradle-check.yml:62]: This seems error prone, ChatGPT suggests the following and I'd recommend we follow this approach

> To achieve an early exit in GitHub Actions workflows without repeating an if condition on every step, you can use a job-level strategy. The most effective method involves splitting your workflow into multiple jobs and using the needs and if conditions at the job level. This way, the execution of subsequent jobs depends on the conditions evaluated in the first job. 

```yaml
jobs:
  check-changes:
    outputs:
      should_continue: ${{ steps.changed-files-specific.outputs.any_changed }}

    steps:
      - uses: actions/checkout@v2

      -id: changed-files-specific
        uses: tj-actions/changed-files@v44
        with:
          files_ignore: |
            release-notes/*.md

  build-and-test:
    needs: check-changes
    if: ${{ needs.check-changes.outputs.should_continue == 'true' }}
    steps:
      - run: echo "Building the project..."
```

**Review by peternied — CHANGES_REQUESTED** [review]
I insist we are very careful with allowing the bypass of the gradle check.

If there are changes that are blocked because of the missing gradle check results, lets revert the initial update so we have more time to iterate on this change

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:36]: Are files under .github related to gradle check in anyway?

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:37]: There are man .md files 
```
ADMINS.md
CHANGELOG-3.0.md
CHANGELOG.md
CODE_OF_CONDUCT.md
CONTRIBUTING.md
DEVELOPER_GUIDE.md
MAINTAINERS.md
README.md
RELEASING.md
SECURITY.md
TESTING.md
TRIAGING.md
```
Do any of these or more runs any kind of validation during gradle check?

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:62]: I did do that before. See https://github.com/opensearch-project/OpenSearch/pull/13498/commits/3389397b2f8aee92443a9f1b45f4a9a7dd49c0d7
However, the required status check at repository level needs check at `job` level and not workflow level.

image

Comment (peternied) [reply .github/workflows/gradle-check.yml:62]: Thanks for the context, bummer it didn't work

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml]: Makes sense. Modified the list

File: .github/workflows/gradle-check.yml
Comment (peternied) [inline .github/workflows/gradle-check.yml:35]: Could you please add a comment after each file/entry that justifies why it should be ignored?

If the answer is we want to go fast, why not just have an admin merge bypassing checks?

Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:35]: I believe saving time and resources is a generic reason even enough for this change? Running a job for 1+ hour and the backport (1+ hour) which adds no value to the run but can be a blocker due to flaky tests. Bypassing is always an option but usually not an opted one.

Also I am not sure if adding a comment would work or not (needs more testing) as I see one more input `files_ignore_separator` in the action used https://github.com/tj-actions/changed-files?tab=readme-ov-file#inputs-%EF%B8%8F

**Comment by gaiksaya** [comment]
@kotwanikunal @reta Can you please review this change? Let me know if I need to add more maintainers.
Thanks!

File: .github/workflows/gradle-check.yml
Comment (dblock) [inline .github/workflows/gradle-check.yml:48]: What's the reason not to include all *.md at root?

File: .github/workflows/gradle-check.yml
Comment (dblock) [inline .github/workflows/gradle-check.yml:51]: Can we split out the gradle check setup and workflow into a separate job, and make the whole job dependent?

File: .github/workflows/gradle-check.yml
Comment (reta) [inline .github/workflows/gradle-check.yml:88]: I think `always()` does not apply anymore:

```suggestion
        if: steps.changed-files-specific.outputs.any_changed == 'true'
```

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:48]: See the conversation here https://github.com/opensearch-project/OpenSearch/pull/13498#discussion_r1586988920

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:51]: See https://github.com/opensearch-project/OpenSearch/pull/13498#discussion_r1586986521

Comment (reta) [reply .github/workflows/gradle-check.yml:51]: :+1: to @dblock 's suggestion, the same condition is propagated to (mostly) every step

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:88]: If `always()` is removed, the step would be skipped if previous check failed as default is `success()`

Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:88]: `failure() && steps.changed-files-specific.outputs.any_changed == 'true'` will be evaluated to false if previous step fails right? In that case we need always() for first condition to be set to true

Comment (reta) [reply .github/workflows/gradle-check.yml:88]: You might be right, thanks @gaiksaya , it seems like semantics of `always()` is more than just `true`

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:51]: So I did try that but does not work as expected. See comment posted above. Let me know if you can think of better way.

File: .github/workflows/gradle-check.yml
Comment (dblock) [reply .github/workflows/gradle-check.yml:36]: They never are, these are github workflows. And the workflow itself runs on pull request target, so it doesn't impact what's used on the PR itself with the changes to it.

File: .github/workflows/gradle-check.yml
Comment (dblock) [reply .github/workflows/gradle-check.yml:51]: I don't think I understand why the gradle check job is different from other jobs. It should be skippable, shouldn't it?

**Review by dblock — COMMENTED** [review]
Looks like @gaiksaya did a ton of iterating on this one and all our suggestions have already been tried :) I am ok with this change as is. One of the concerns I do have is that a gradle check will visibly pass even if it did nothing.

Will leave for more comments for now.

File: .github/workflows/gradle-check.yml
Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:51]: The issue is this repository has added `gradle-check` as a required workflow check before merging any PR. GitHub is unable to resolve the conflict if the workflow is skipped due to any condition. It HAS to be passed. Not failed, not skipped :(

**Comment by reta** [comment]
@gaiksaya @dblock have some thoughts regarding the problem over the weekend, I think the idea is sound but the way we approach it probably not. Primarily, because  for some reason, we only focus on Gradle check but the same "skip me" rules apply to other workflows like `assemble` (and very likely even `precommit`). 

What if we combine changeset detection + Gradle build instead, and just skip most of the tasks if there are no relevant changes?

**Comment by dblock** [comment]
> What if we combine changeset detection + Gradle build instead, and just skip most of the tasks if there are no relevant changes?

Maybe.

I am confused why we needed to do anything special about gradle check in terms of requiring it to pass. Isn't that what any check does by default?

**Comment by reta** [comment]
> I am confused why we needed to do anything special about gradle check in terms of requiring it to pass. Isn't that what any check does by default?

I mean - we don't need to run the whole test suite if only CHANGELOG.md changes, right?

**Comment by gaiksaya** [comment]
> > What if we combine changeset detection + Gradle build instead, and just skip most of the tasks if there are no relevant changes?
> 
> Maybe.
> 
> I am confused why we needed to do anything special about gradle check in terms of requiring it to pass. Isn't that what any check does by default?


Other workflows are not set to mandatory at branch settings. Current settings for this repo that applies to all branches:
image

**Comment by gaiksaya** [comment]
> @gaiksaya @dblock have some thoughts regarding the problem over the weekend, I think the idea is sound but the way we approach it probably not. Primarily, because for some reason, we only focus on Gradle check but the same "skip me" rules apply to other workflows like `assemble` (and very likely even `precommit`).
> 
> What if we combine changeset detection + Gradle build instead, and just skip most of the tasks if there are no relevant changes?

I believe main reason would be other workflows are not as flaky and not so time consuming. Gradle check is a whole different story. What you are suggesting does not need a common workflow. Just adding `path-ignore` in those workflows would work as GH would just skip them and won't wait at PR merge asking for status. 
Also, I believe you mean gradle tasks here? `just skip most of the tasks ` That would need to break down the gradle check right?

**Comment by reta** [comment]
Thanks @gaiksaya 

> Drawback: Each workflow would have it own list to ignore which I think is fine?

I think the ignore list is per repository, right? The only thing we may need to figure out - what has changed (changeset)

> Also, I believe you mean gradle tasks here? just skip most of the tasks That would need to break down the gradle check right?

Correct, more like `onlyIf { hasChanges() }`  condition.

**Comment by reta** [comment]
> It's per workflow. Example: https://github.com/opensearch-project/OpenSearch/pull/13477/files# and [path.ignore](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-excluding-paths)

This is how we implemented it, correct, but it does not need to be

> Can this be a part of gradle check breakdown issue?

May be, I just wanted to run the idea over you folks, I haven't prototyped it

**Comment by dblock** [comment]
I think we should decide whether we want to merge this as is. WDYT @gaiksaya @reta? I am good with whatever you two choose!

**Comment by gaiksaya** [comment]
> I think we should decide whether we want to merge this as is. WDYT @gaiksaya @reta? I am good with whatever you two choose!

I am good to merge as is too.

**Comment by reta** [comment]
> I am good to merge as is too.

We have  no better **working** option at the moment (and @gaiksaya did really good job here figuring out GA internals)  but do we need to rush the merge? The workflow becomes difficult to follow and change, this is the concern I have (and that is the motivation to look for other options).

File: .github/workflows/gradle-check.yml
Comment (andrross) [reply .github/workflows/gradle-check.yml:51]: @gaiksaya I think there might be another approach as described here: https://stackoverflow.com/questions/65844033/conditional-needs-in-github-actions

We'd need to something like:

```
jobs:
  changed-files-specific:
    steps:
      - set the any_changed boolean

  gradle-check:
    runs-on: ubuntu-latest
    needs: changed-files-specific
    if: ${{ needs.changed-files-specific.outputs.any_changed == 'true' }}
    steps:
      - run the gradle check steps

  check-result:
    needs: gradle-check
    if: always()
    steps:
      - name: fail if conditional job failed
        if: ${{ needs.changed-files-specific.outputs.any_changed == 'true' && needs.gradle-check.result == 'failure' }}
        run: exit 1
```

And then the branch protection rule would be on `check-result` which will always run but is smart enough to ignore gradle-check if no relevant files changed. Would this work?

Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:51]: Sorry for delayed response. Missed this. Not really, it has to pass if it is required. Cannot skip or fail. So I believe only approach I think is follow what you just posted and add one more step which will print something like "Skipping check" and show it as pass. Let me know if everyone are okay with this approach

Comment (andrross) [reply .github/workflows/gradle-check.yml:51]: @gaiksaya In my suggestion I think the `check-result` step will always run, and it is the thing that will be marked as required. Why would you need another step?

Comment (gaiksaya) [reply .github/workflows/gradle-check.yml:51]: I think you are right. @reta @dblock Are you okay with this approach? I can go ahead and test it out and move to implementation then.

Comment (reta) [reply .github/workflows/gradle-check.yml:51]: 👍 thanks @gaiksaya !

Comment (dblock) [reply .github/workflows/gradle-check.yml:51]: Yes i am, thank you.

Comment (gaiksaya) [reply .github/workflows/gradle-check.yml]: Updated with code with suggested changes. Thanks! Also works as expected. Here is the sample: https://github.com/gaiksaya/OpenSearch/pull/12

**Review by reta — APPROVED** [review]

**Review by peternied — APPROVED** [review]
Looks like we might need to update the required check after this is merged, @gaiksaya can you watch/resolve any issues that come up post merge?

**Comment by gaiksaya** [comment]
Sorry @peternied @reta Need approval again in order for minimal approval count workflow to pass. I will be merging the PR after long weekend to be available for surprise failures :)

**Comment by gaiksaya** [comment]
I added backporting to 2.x label. Any more branches that this PR should be backported to? Changing the required workflow to pass check from `gradle-check` to `check-result`

**Comment by reta** [comment]
> I added backporting to 2.x label. Any more branches that this PR should be backported to?

I think 1.x / 1.3 would benefit from that as well, thank you.

**Comment by gaiksaya** [comment]
> > I added backporting to 2.x label. Any more branches that this PR should be backported to?
> 
> I think 1.x / 1.3 would benefit from that as well, thank you.

Done! Thank you!

**Pull request merged**
